File: pa2/Lex.c
//Rahul Patel
//rpate3
//pa2

#include<stdio.h>
#include<string.h>
#include<stdlib.h> 

#include "List.h"

#define MAX 200

//Insertsort operation
List insertsort(char** array, int length){


	List list = newList();
	append(list, 0);//append 0 onto the list
	for(int i = 1; i <length; i++){
		char *temp = array[i]; 
		int index = i-1; 
		moveTo(list, index); //moves cursor to correct position
		while(index>-1 && strcmp(temp, array[getElement(list)]) <0){//sorts lex
			index--; 
			movePrev(list); 
		}
		if(getIndex(list)==-1){ //condition for null cursor
			prepend(list, i); 
		}else{
			insertAfter(list, i); 
		}
	}
	return list;
}


//reads in the input file and stores them in an array of strings
char** read(FILE* input, int lines){
	char** Sarray= malloc(sizeof(char**)*lines); 
	char line[200]; 
	for(int i=0; i<lines; i++){
		fgets(line, 200, input);
		Sarray[i] = strdup(line); 
	}
	return Sarray; 
}

void freed(char** lines, int linec, List list){
	for(int i=0; i<linec; i++){
		free(lines[i]); 		
	}
	free(lines); 
	freeList(&list);
}








int main(int argc,char* argv[]){
	if(argc > 3){  //Checks command line
		printf("Invalid inputs\n"); 
		exit(1);
	}

	//read in the files
	FILE* input = fopen(argv[1],"r");
	FILE* output = fopen(argv[2], "w");


	int linecount = 0; 
	char lines[MAX]; 
	
	//goes through the file and counts the lines
	while(fgets(lines, MAX, input)!= NULL) linecount++; 


	fclose(input); 
	input = fopen(argv[1], "r"); 
	char** inlines = read(input, linecount); 
	
	//carries out insertsort and stores result into a list
 	List list = insertsort(inlines, linecount); 

	//prints out list by going through the indicies
	for(moveTo(list,0); getIndex(list)>=0; moveNext(list)){
		printf( "%s", inlines[getElement(list)]); 
	}
	
	//close files
	fclose(input); 
	fclose(output);
	freed(inlines, linecount, list);  
	return(0); 
}

File: pa1/Lex.java
//Rahul Patel
//UserID: rpate3
//pa1

import java.io.*;
import java.lang.System;
import java.util.Scanner;

class Lex{

	
	public static void main(String[] args){
		if(args.length < 2){				//Checks if command line is valid
			System.err.println("Invalid inputs");
			System.exit(1); 		
	} 
	//Allocated space for file being read in and one being outputted with exception handling
	Scanner in = open(args[0]);
	PrintWriter out = null;	
	try{
		out = new PrintWriter(new FileWriter(args[1]) );
	} catch (FileNotFoundException e){
		System.err.println("File can not be created"); 
		System.exit(1); 
	} catch(IOException e){
		System.err.println("File can not be created");
		System.exit(1);
	}	
	int linec = countlines(in); //Number of lines n in file arg[0]
	in = open(args[0]);
	String [] array = read(in, linec);
	List l = insertsort(array);//A list with the sorted items	
	for(l.moveTo(0); l.getIndex()>=0; l.moveNext()){//Print out the elements of indicies
		out.println(array[l.getElement()]);
	}
	out.close();
}
	//Used to access/open the .txt file
	static Scanner open(String filename){ 
		Scanner in = null; 
		try{
			in = new Scanner(new File(filename)); 
		} catch (IOException e){ 
			System.err.println("Invalid file");
			System.exit(1); 
		}
		return in; 
	}

	//keeps track of the number of lines in the file 
	static int countlines(Scanner file){ 
		int i = 0; 
		for(i=i; file.hasNextLine(); file.nextLine()) i++; 
		return i; 
	}
	
	//goes through each line and stores it into a String array
	static String[] read(Scanner file, int lines){
		String[] Stringarray = new String[lines]; 
		for(int i=0; file.hasNextLine(); i++){
		Stringarray[i] = file.nextLine();  
		}
		return Stringarray; 
		
	}
	
	//sorts the list
	static List insertsort(String[] file){
		List list = new List(); //creates new list  
		list.append(0);//appends a 0
		for(int j =1; j< file.length; j++){ 
			String hold = file[j]; 
			int cursor = j-1; 
			list.moveTo(cursor); //sets cursor to current position
		
		while(cursor > -1 && hold.compareTo(file[list.getElement()])<0){//while loop to move the cursor along the list to find the proper position for insertion
				cursor--; 
				list.movePrev(); 
			}
		if(list.getIndex() == -1){ //if cursor = null, prepend
			list.prepend(j); 
		}else{
			list.insertAfter(j); 
		
			}	
			}
	return list; 
	}

}
